add verifying signed message, using pubkey recovery

rubensayshi opened the pull request:
**Work in Progress**

I ported the code for pubkey recovery from bitcoin core using ctypes, the same way as the other code is written.

The `VerifyMessage` (and `BitcoinMessage` class) are still in the `test_signverify_message.py` because I have no idea where you would want these to go? please let me know where you want me to put them

I'm not entirely sure about the `CPubKey.recover_compact`, as a python dev I'd say it should be a `@classmethod` (like it is now) but to match the bitcoin core code (which is what you've been doing) should it be different?

tested against `py2.7` and `py3.4`, to test it against `libeay` I have to run it on windows I guess, or is it 100% certain that everything 'just works'?

PS. is there any particular reason for not having travis-ci setup to run the tests on the various python versions?

rubensayshi commented:
ugh I'm getting random failures when trying to do the signing (https://github.com/blocktrail/python-bitcoinlib/tree/sign-verify-message2)

rubensayshi commented:
okay, seem to have it working now except I keep running into some odd thing where if `_ssl.BN_bin2bn` stops when it runs into `x00`, eg;
```b'\x00\xeb\xbb\x00\x17\x0c\x93(\xd4U\xfd\xc7\xfe\x1d\xf9&\xf9\xff\x1e\x8dB\xc8\x98~\xe1Y\x8e3g\xe4\x88\x0c\x90'``` becomes `b'\xeb\xbb'`
and
`b'\x00\xc3\xd7!7\x01\x99\xcd%+\xef\xcd+n\xac\x00\x0b\x82\xdb\xb72+\xb4\xe6]\x80\xeb\xc0s\x99\x94"\xe3'` becomes `b'\xc3\xd7!7\x01\x99\xcd%+\xef\xcd+n\xac`

Maybe @petertodd you can help me with this because tbh I this is the first time for be using this stuff

PS. should I make a `DERSignature(ImmutableSerializable)` class or just use the code from `IsLowDERSignature` to split decode the signature?

petertodd commented:
FWIW, I'm travelling right now and am pretty busy until the MIT Bitcoin Expo is over; I'll try to get to this when I get back after that.

Thanks!

rubensayshi commented:
@petertodd I fixed the issue with the null bytes, if you could tell me where (which files) you'd prefer these things to go then this stuff should be ready:

 - `class BitcoinMessage(ImmutableSerializable)`
 - `def SignMessage(key, message)`
 - `def VerifyMessage(address, message, sig)`

petertodd commented:
So bitcoin signatures aren't consensus critical, so rather than putting them in bitcoin.core.signature, let's put that functionality in bitcoin.signature instead. As for the code you added to bitcoin.core.key though... meh, leave it there for now; this can be cleaned up later.

rubensayshi commented:
@petertodd I've been slacking lately, sorry for that, I want / need to pick up this PR ...

I'll move `bitcoin.core.signature` to `bitcoin.signature`, the classes I listed before are currently in `test_signverify_message.py`, normally I'd move those into `bitcoin.messages` but ... that's already being used and that has a completely different context, any idea where these should go?

 - `class BitcoinMessage(ImmutableSerializable)`
 - `def SignMessage(key, message)`
 - `def VerifyMessage(address, message, sig)`

rubensayshi commented:
I just moved them into `bitcoin.sign_verify_message`, just to already seperate them, but it should get a more sensible name, I just can't come up with anything good ...

petertodd commented:
How about just bitcoin.signmessage?

It'd be good to have signmessage.py and verifymessage.py scripts in the examples/ directory showing the functionality off too.

rubensayshi commented:
okay, `bitcoin.signmessage` sounds good, moved!  
also added example `signmessage.py` that does both, not really neccesary to have 2 scripts.  

and I rebase and squashed all the commits into 1

rubensayshi commented:
@petertodd I think this is good to go?

petertodd commented:
Thanks!

Sorry, was out of the country, as usual. :/

rubensayshi commented:
awesome! thanks!

can we also get a v0.4.1 ;) ?

petertodd commented:
Time for a v0.5.0 I think.
